dbms/src/Storages/Page/V3/PageStorageImpl.cpp

```diff
@@ -269,13 +279,135 @@ void PageStorageImpl::traverseImpl(const std::function<void(const DB::Page & pag
 {
     // If another thread is running gc, just return;
     bool v = false;
     if (!gc_is_running.compare_exchange_strong(v, true))
         return false;
 
+    const GCTimeStatistics statistics = doGC(write_limiter, read_limiter);
+    assert(statistics.stage != GCStageType::Unknown); // `doGC` must set the stage
+    LOG_DEBUG(log, statistics.toLogging());
+
+    return statistics.executeNextImmediately();
+}
+
+// Remove external pages for all tables
+// TODO: `clean_external_page` for all tables may slow down the whole gc process when there are lots of table.
+void PageStorageImpl::cleanExternalPage(Stopwatch & gc_watch, GCTimeStatistics & statistics)
+{
+    // Fine grained lock on `callbacks_mutex`.
+    // So that adding/removing a storage will not be blocked for the whole
+    // processing time of `cleanExternalPage`.
+    std::shared_ptr<ExternalPageCallbacks> ns_callbacks;
+    {
+        std::scoped_lock lock{callbacks_mutex};
+        // check and get the begin iter
+        statistics.num_external_callbacks = callbacks_container.size();
+        auto iter = callbacks_container.begin();
+        if (iter == callbacks_container.end()) // empty
+        {
+            statistics.clean_external_page_ms = gc_watch.elapsedMillisecondsFromLastTime();
+            return;
+        }
+
+        assert(iter != callbacks_container.end()); // early exit in the previous code
+        // keep the shared_ptr so that erasing ns_id from PageStorage won't invalid the `ns_callbacks`
+        ns_callbacks = iter->second;
+    }
+
+    Stopwatch external_watch;
+    // get all alive external ids from all namespaces
+    auto alive_external_ids = page_directory->getAliveExternalIds();
+    statistics.external_page_get_alive_ns = external_watch.elapsedFromLastTime();
+
+    SYNC_FOR("before_PageStorageImpl::cleanExternalPage_execute_callbacks");
+
+    while (true)
+    {
+        // Assume calling the callbacks after erasing ns_is is safe.
+
+        // the external pages on disks.
+        auto pending_external_pages = ns_callbacks->scanner();
+        statistics.external_page_scan_ns += external_watch.elapsedFromLastTime();
+        // If we can not find ns_id in `alive_external_ids`, there could be two situation:
+        //   1. all external ids of this `ns_id` have been removed from `page_directory`,
+        //      the DeltaMergeStore should be physically dropped and clean the external
+        //      pages (DTFile).
+        //   2. the `ns_id` is concurrently added after we get `alive_external_ids`, we
+        //      should not remove the external pages on disk.
+        // TODO: Is it ok?
+        if (auto iter = alive_external_ids.find(ns_callbacks->ns_id); iter != alive_external_ids.end())
+        {
+            // remove the external pages that is not alive now.
+            ns_callbacks->remover(pending_external_pages, iter->second);
+        }
+        // Else situation 2 mentioned above, just skip `remover`.
```

Comment:
/hold

There is a corner case that these lines lead to removing DTFile being used. We must ensure to execute `scanner`
before `getAliveExternalIds`
And that changes will make this PR too complicated. I will split the changes into another PR first.